equihash/equihash.go

```diff
@@ -0,0 +1,722 @@
+package consensus
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"hash"
+	"math"
+	"math/big"
+	"reflect"
+	"sort"
+
+	blake2b "github.com/minio/blake2b-simd"
+)
+
+const (
+	// size of word
+	wordSize = 32
+	wordMask = (1 << wordSize) - 1
+	byteMask = 0xFF
+	// size of each hash key
+	hashSize = 64
+	// N is the number of hash digests used to find a mining solution
+	N = 96
+	// K is the exponent for xor'ing 2^k hash digests for solution
+	K = 5
+)
+
+var (
+	errBadArg           = errors.New("invalid argument")
+	errWriteLen         = errors.New("didn't write full len")
+	errKLarge           = errors.New("k should be less than n")
+	errCollisionLen     = errors.New("collision length too big")
+	errSmallBitLen      = errors.New("bitLen < 8")
+	errSmallWordSize    = errors.New("wordSize < 7+bitLen")
+	errBadOutLen        = errors.New("outLen != 8*outWidth*len(in)/bitLen")
+	errDuplicateIndices = errors.New("duplicate indices")
+	errPairWiseOrdering = errors.New("bad pair-wise ordering")
+	errBadWord          = errors.New("bad word")
+	exccPrefix          = "excc"
+	bigZero             = big.NewInt(0)
```

Comment:
Why cant `exccPrefix` and `bigZero` be const?

Reply:
Only numeric types, strings and bools can be constants... so `bigZero` cannot be constant but `exccPrefix` can. I will refactor to make `exccPrefix` a constant.